cosmwasm/packages/wasmi-runtime/src/recursion_depth.rs

```diff
@@ -0,0 +1,34 @@
+use std::sync::SgxMutex;
+
+use lazy_static::lazy_static;
+
+use enclave_ffi_types::EnclaveError;
+
+const RECURSION_LIMIT: i32 = 5;
+
+lazy_static! {
+    /// This counter tracks the recursion depth of queries,
+    /// and effectively the amount of loaded instances of WASMI.
+    ///
+    /// It is incremented before each computation begins and is decremented after each computation ends.
+    static ref RECURSION_DEPTH: SgxMutex<i32> = SgxMutex::new(0);
+}
+
+pub fn increment() -> Result<(), EnclaveError> {
+    let mut depth = RECURSION_DEPTH.lock().unwrap();
+    if *depth == RECURSION_LIMIT {
+        return Err(EnclaveError::ExceededRecursionLimit);
+    }
+    *depth += 1;
+    Ok(())
+}
+
+pub fn decrement() {
+    let mut depth = RECURSION_DEPTH.lock().unwrap();
+    *depth -= 1;
+}
```

Comment:
I think this effectively allows 6 VMs in parallel, right?
I'd feel more comfortable if we only allow 5 VMs in parallel and keep the memory at 128 MiB.

cc @toml01 @Cashmaney

Reply:
I.e set `RECURSION_LIMIT = 4`.

Reply:
No. This counter also includes the initial call into the enclave, so a recursion limit of 2 means you can call `query_chain` once. As you've seen, I increment the counter in the `init` and `handle` entry points as well.
It effectively tracks how many running instances of wasmi we have.

Reply:
Okay so let's say I'm calling and endless loop of sub query.

- 1st call: 0 VMs total => `depth (0) != RECURSION_LIMIT (5)` => `depth <- 1`
- 2nd call: 1 VMs total => `depth (1) != RECURSION_LIMIT (5)` => `depth <- 2`
- 3th call: 2 VMs total => `depth (2) != RECURSION_LIMIT (5)` => `depth <- 3`
- 4th call: 3 VMs total => `depth (3) != RECURSION_LIMIT (5)` => `depth <- 4`
- 5th call: 4 VMs total => `depth (4) != RECURSION_LIMIT (5)` => `depth <- 5`
- 6th call: 5 VMs total => `depth (5) == RECURSION_LIMIT (5)` => `error`

Cool, you're right.